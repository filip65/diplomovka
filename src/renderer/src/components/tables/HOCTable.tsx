import { useAppDispatch, useAppSelector } from '../../store/hooks';
import PDFTable from './PDFTable';
import Title from '../Title';
import {
  Autocomplete,
  Box,
  Grid,
  Paper,
  SortDirection,
  styled,
  TextField,
} from '@mui/material';
import {
  ActionCellBottom,
  ActionCellBottomRight,
  ActionCellRight,
  DataTable,
  TableBody,
  TableCell,
  TableRow,
} from './Table';
import TableActionButton from './TableActionButton';
import { useError } from '../providers/ErrorProvider';
import { CellValue, CellType } from '@renderer/store/rootReducer';
import { RootSelectors } from '@renderer/store/store';
import React, { useEffect, useRef } from 'react';

const TableWrapper = styled(Box)`
  overflow-x: auto;
`;

export default function withTable(
  TableInput: React.FC<any>,
  TableDataHeader: React.FC<any>,
  selectors: RootSelectors,
  actions: any,
) {
  return () => {
    const dispatch = useAppDispatch();
    const { openError } = useError();

    const data = useAppSelector(selectors.data);
    const dynCols = useAppSelector(selectors.dynCols);
    const values = useAppSelector(selectors.values);

    const tableContentRef = useRef<HTMLDivElement>(null);
    const prevDataLength = useRef<number>(data[0]?.length || 0);

    const deleteColumn = (col: number) => {
      dispatch(actions.deleteColumn(col));
    };

    const sortByYear = (sortDirection: SortDirection) => {
      try {
        dispatch(actions.sortTableByYear(sortDirection));
      } catch (error: any) {
        openError(error);
      }
    };

    const sortByItemNumber = (sortDirection: SortDirection) => {
      dispatch(actions.sortTableByItemNumber(sortDirection));
    };

    useEffect(() => {
      const currentDataLength = data[0]?.length || 0;

      if (currentDataLength > prevDataLength.current) {
        if (tableContentRef.current) {
          tableContentRef.current.scrollLeft =
            tableContentRef.current.scrollWidth;
        }
      }

      prevDataLength.current = currentDataLength;
    }, [data]);

    return (
      <>
        <Title
          onSortYear={actions.sortTableByYear ? sortByYear : undefined}
          onSortItems={
            actions.sortTableByItemNumber ? sortByItemNumber : undefined
          }
        />
        <Paper className="hideInPrint">
          <Grid container>
            <Grid xs={4} item>
              <TableInput selectors={selectors} actions={actions} />
            </Grid>
            <Grid xs={8} item>
              <TableWrapper ref={tableContentRef}>
                <DataTable>
                  <TableDataHeader selectors={selectors} actions={actions} />
                  <TableBody>
                    {data.map((rowData, row) => {
                      return (
                        <Row
                          key={values[row].id}
                          row={row}
                          data={rowData}
                          selectors={selectors}
                          actions={actions}
                        />
                      );
                    })}
                    <TableRow>
                      {data[0].map((_value, col) => {
                        return (
                          <ActionCellBottom key={col}>
                            {dynCols && (
                              <TableActionButton
                                buttonType="delete"
                                onClick={() => deleteColumn(col)}
                              />
                            )}
                          </ActionCellBottom>
                        );
                      })}
                      <ActionCellBottomRight></ActionCellBottomRight>
                    </TableRow>
                  </TableBody>
                </DataTable>
              </TableWrapper>
            </Grid>
          </Grid>
        </Paper>

        <PDFTable selectors={selectors} />
      </>
    );
  };
}

type RowProps = {
  data: CellValue[];
  row: number;
  selectors: RootSelectors;
  actions: any;
};

const Row: React.FC<RowProps> = React.memo(
  ({ data, row, selectors, actions }) => {
    const dispatch = useAppDispatch();

    const dynRows = useAppSelector(selectors.dynRows);
    const headers = useAppSelector(selectors.headers);

    const deleteRow = (row: number) => {
      dispatch(actions.deleteRow(row));
    };

    return (
      <TableRow>
        {data.map((_, col) => (
          <Cell
            key={headers[col].id}
            actions={actions}
            selectors={selectors}
            col={col}
            row={row}
          />
        ))}
        {dynRows && (
          <ActionCellRight>
            <TableActionButton
              buttonType="delete"
              onClick={() => deleteRow(row)}
            />
          </ActionCellRight>
        )}
      </TableRow>
    );
  },
);

type CellProps = {
  row: number;
  col: number;
  selectors: RootSelectors;
  actions: any;
};

const Cell: React.FC<CellProps> = React.memo(
  ({ row, col, selectors, actions }) => {
    const dispatch = useAppDispatch();

    const header = useAppSelector(selectors.selectHeaderByIndex(col));
    const value = useAppSelector(selectors.selectDataByPosition(row, col));
    const rowType = useAppSelector(selectors.getRowType(row));

    const handleChangeData = function (value: string) {
      // zmena , na . pri cislach, lebo inak sa hodnota zmeni na 0
      if (rowType === CellType.NUMBER) {
        value = value.replace(',', '.');

        // zakazanie zadat viac . alebo , v cisle
        if (value.split('.').length > 2) {
          return;
        }

        // zakazanie zadat pismeno v cisle
        if (/[a-zA-Z]/.test(value)) {
          return;
        }

        if (!/0+$/.test(value)) {
          // remove unnecessary 0 from beginning of the number
          if (/^\d+\.\d+$/.test(value)) {
            value = parseFloat(value).toString();
          }

          // zabranenie zadat viacero 0 na zaciatok cisla
          if (/^0+/.test(value) && !/^0+\./.test(value)) {
            value = parseFloat(value).toString();
          }
        }
      }

      dispatch(
        actions.setDataOnIndex({
          data: value,
          row,
          col,
          type: rowType,
        }),
      );
    };

    return (
      <TableCell>
        <TextField
          placeholder={rowType === CellType.NUMBER ? '0' : ''}
          value={value}
          onChange={(e) => handleChangeData(e.target.value ?? '')}
          onWheel={(event) => event.currentTarget.blur()}
          sx={{
            position: 'absolute',
            inset: 0,

            '& .MuiOutlinedInput-root': {
              borderRadius: '0',
              '& fieldset': {
                border: 'none',
              },
            },

            input: {
              height: '48px',
              padding: 0,
              textAlign: 'center',
            },
          }}
        />
        {/* Toto sa odstranovalo ale nechavam to tu, keby sa to chcelo v buducnosti pridat */}
        {/* {rowType === CellType.NUMBER || rowType === CellType.STRING ? (
          
        ) : (
          <Autocomplete
            value={header.options?.find(
              (option) => option.value === value.toString(),
            )}
            options={header.options ?? []}
            getOptionLabel={(option) => option.label}
            renderInput={(params) => <TextField {...params} />}
            clearIcon={false}
            onChange={(_, value) => {
              handleChangeData(value?.value ?? '');
            }}
            sx={{
              position: 'absolute',
              inset: 0,

              '.MuiOutlinedInput-root': {
                borderRadius: '0',
                height: '48px',
                paddingRight: '36px !important',

                fieldset: {
                  border: 'none',
                  height: '48px',
                },
              },

              input: {
                height: `${48 - 18}px !important`,
                padding: '0px !important',
              },
            }}
          />
        )} */}
      </TableCell>
    );
  },
);
