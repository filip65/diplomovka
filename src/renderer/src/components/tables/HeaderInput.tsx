import { RootSelectors } from '@renderer/store/store';
import { useAppDispatch, useAppSelector } from '../../store/hooks';
import { ActionCellRight, TableCell, TableHead, TableRow } from './Table';
import TableActionButton from './TableActionButton';
import TableInput from './TableInput';
import React, { forwardRef, memo, useEffect, useRef } from 'react';

type HeaderInputProps = {
  selectors: RootSelectors;
  actions: any;
};

const HeaderInput: React.FC<HeaderInputProps> = ({ selectors, actions }) => {
  const dispatch = useAppDispatch();
  const headers = useAppSelector(selectors.headers);
  const dynCols = useAppSelector(selectors.dynCols);

  const inputRef = useRef<HTMLInputElement>(null);
  const prevHeadersLength = useRef<number>(headers?.length || 0);

  const addColumn = () => {
    dispatch(actions.addColumn());
  };

  const disableAddButton = headers.some((header) => !header.label);

  useEffect(() => {
    const currentHeadersLength = headers.length || 0;

    if (currentHeadersLength > prevHeadersLength.current) {
      inputRef.current?.focus();
    }

    prevHeadersLength.current = currentHeadersLength;
  }, [headers]);

  return (
    <TableHead>
      <TableRow>
        {headers.map((header, index) => (
          <Input
            ref={index === headers.length - 1 ? inputRef : null}
            key={header.id}
            actions={actions}
            index={index}
            selectors={selectors}
          />
        ))}

        {dynCols && (
          <ActionCellRight $topBorder={false}>
            <TableActionButton
              buttonType="add"
              onClick={addColumn}
              disabled={disableAddButton}
            />
          </ActionCellRight>
        )}
      </TableRow>
    </TableHead>
  );
};

export default HeaderInput;

type InputProps = {
  selectors: RootSelectors;
  actions: any;
  index: number;
  ref: any;
};

const Input: React.FC<InputProps> = memo(
  forwardRef(({ selectors, actions, index }, ref) => {
    const dispatch = useAppDispatch();

    const header = useAppSelector(selectors.selectHeaderByIndex(index));

    const handleChangeHeader = (value: string, index: number) => {
      dispatch(
        actions.setHeaderOnIndex({
          index,
          data: {
            id: header.id,
            value,
          },
        }),
      );
    };

    return (
      <TableCell>
        <TableInput
          ref={ref as any}
          defaultValue={header.label}
          onChange={(e) => handleChangeHeader(e.target.value, index)}
        />
      </TableCell>
    );
  }),
);
