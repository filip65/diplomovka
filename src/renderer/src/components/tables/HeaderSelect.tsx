import { RootSelectors } from '@renderer/store/store';
import { useAppDispatch, useAppSelector } from '../../store/hooks';
import { ActionCellRight, TableCell, TableHead, TableRow } from './Table';
import TableActionButton from './TableActionButton';
import TableSelect from './TableSelect';
import React from 'react';

const selectCol = [
  { value: 7, label: 'Priamy materiál' },
  { value: 8, label: 'Priame mzdy' },
  { value: 9, label: 'Ostatné priame náklady' },
  { value: 10, label: 'Výrobná réžia technologická' },
  { value: 11, label: 'Výrobná réžia všeobecná' },
  { value: 12, label: 'Správna réžia' },
  { value: 13, label: 'Odbytová réžia' },
  { value: 14, label: 'Zásobovacia réžia' },
  { value: 15, label: 'Dopravná réžia' },
];

type HeaderSelectProps = {
  selectors: RootSelectors;
  actions: any;
};

const HeaderSelect: React.FC<HeaderSelectProps> = ({ selectors, actions }) => {
  const headers = useAppSelector(selectors.headers);

  const dispatch = useAppDispatch();

  const addColumn = () => {
    dispatch(actions.addColumn());
  };

  const disableAddButton = headers.some((value) => !value.label);

  return (
    <TableHead>
      <TableRow>
        {headers.map((header, index) => (
          <Select
            key={header.id}
            selectors={selectors}
            actions={actions}
            index={index}
          />
        ))}

        {headers.length < selectCol.length && (
          <ActionCellRight $topBorder={false}>
            <TableActionButton
              buttonType="add"
              onClick={addColumn}
              disabled={disableAddButton}
            />
          </ActionCellRight>
        )}
      </TableRow>
    </TableHead>
  );
};

export default HeaderSelect;

type SelectProps = {
  selectors: RootSelectors;
  actions: any;
  index: number;
};

const Select: React.FC<SelectProps> = React.memo(
  ({ selectors, actions, index }) => {
    const dispatch = useAppDispatch();

    const header = useAppSelector(selectors.selectHeaderByIndex(index));
    const headers = useAppSelector(selectors.headers);

    const handleChangeHeader = function (
      event: React.ChangeEvent<HTMLSelectElement>,
    ) {
      dispatch(
        actions.setHeaderOnIndex({
          index,
          data: {
            id: header.id,
            value: event.target.value,
          },
        }),
      );
    };

    let availableHeadersOptions = selectCol
      .filter(
        (option: { value: number; label: string }) =>
          !headers.some((header) => header.label === option.label),
      )
      .map((option: any) => option.label);

    return (
      <TableCell>
        <TableSelect
          value={header.label}
          onChange={(e) => handleChangeHeader(e as any)}
          options={[...availableHeadersOptions, header.label].filter(
            (option) => !!option,
          )}
        />
      </TableCell>
    );
  },
);
