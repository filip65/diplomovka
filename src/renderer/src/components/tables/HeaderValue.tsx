import { Typography } from '@mui/material';
import { TableCell, TableHead, TableRow } from './Table';
import { RootSelectors } from '@renderer/store/store';
import { useAppSelector } from '@renderer/store/hooks';

type HeaderValueProps = {
  selectors: RootSelectors;
};

const HeaderValue: React.FC<HeaderValueProps> = ({ selectors }) => {
  const headers = useAppSelector(selectors.headers);

  return (
    <TableHead>
      <TableRow>
        {headers.map((header) => (
          <TableCell
            key={header.id}
            sx={{
              minWidth: '12.75vw',
              maxWidth: '12.75vw',
              width: '12.75vw',
            }}
          >
            <Typography
              sx={{
                textAlign: 'center',
                fontSize: '14px',
                fontWeight: 'bold',
              }}
              dangerouslySetInnerHTML={{ __html: header.label }}
            />
          </TableCell>
        ))}
      </TableRow>
    </TableHead>
  );
};

export default HeaderValue;
