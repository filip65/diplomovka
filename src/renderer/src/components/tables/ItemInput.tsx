import React from 'react';
import { useAppDispatch, useAppSelector } from '../../store/hooks';
import { RootSelectors } from '../../store/store';
import {
  ActionCellBottom,
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableRow,
} from './Table';
import TableActionButton from './TableActionButton';
import TableInput from './TableInput';

type ItemInputProps = {
  selectors: RootSelectors;
  actions: any;
};

const ItemInput: React.FC<ItemInputProps> = ({ selectors, actions }) => {
  const dispatch = useAppDispatch();

  const corner = useAppSelector(selectors.corner);
  const values = useAppSelector(selectors.values);
  const dynRows = useAppSelector(selectors.dynRows);

  const addRow = () => {
    dispatch(actions.addRow());
  };

  const disableAddRow = values.some((value) => !value.value);

  return (
    <Table>
      <TableHead>
        <TableRow>
          <TableCell>{corner}</TableCell>
        </TableRow>
      </TableHead>
      <TableBody>
        {values.map((value, index) => {
          return (
            <Input
              key={value.id}
              selectors={selectors}
              actions={actions}
              index={index}
            />
          );
        })}

        {dynRows && (
          <TableRow>
            <ActionCellBottom>
              <TableActionButton
                buttonType="add"
                onClick={addRow}
                disabled={disableAddRow}
              />
            </ActionCellBottom>
          </TableRow>
        )}
      </TableBody>
    </Table>
  );
};

export default ItemInput;

type InputProps = {
  selectors: RootSelectors;
  actions: any;
  index: number;
};

const Input: React.FC<InputProps> = React.memo(
  ({ selectors, actions, index }) => {
    const dispatch = useAppDispatch();

    const value = useAppSelector(selectors.selectValueByIndex(index));

    const handleChangeInput = function (e: any) {
      dispatch(actions.setItemsOnIndex({ data: e.label || e.value, index }));
      dispatch(
        actions.setValuesOnIndex({
          index,
          data: {
            id: value.id,
            value: e.value,
          },
        }),
      );
    };

    return (
      <TableRow>
        <TableCell>
          <TableInput
            defaultValue={value.value}
            onChange={(e) => handleChangeInput(e.target)}
            inputTextAlign="left"
          />
        </TableCell>
      </TableRow>
    );
  },
);
