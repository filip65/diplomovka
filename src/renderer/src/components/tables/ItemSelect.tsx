import { useAppDispatch, useAppSelector } from '../../store/hooks';
import { RootSelectors } from '../../store/store';
import {
  Table,
  TableBody,
  TableCell,
  TableCorner,
  TableHead,
  TableRow,
} from './Table';
import { Autocomplete, styled, TextField } from '@mui/material';
import TableInput from './TableInput';
import TableActionButton from './TableActionButton';
import { ADD_CUSTOM_ITEM_LABEL } from '@renderer/chartOfAccounts';
import React from 'react';
import { ItemSelectOption } from '@renderer/store/rootReducer';

const AddCell = styled(TableCell)`
  position: relative;
  height: 28px;
`;

type ItemSelectProps = {
  selectors: RootSelectors;
  actions: any;
};

const ItemSelect: React.FC<ItemSelectProps> = ({ selectors, actions }) => {
  const corner = useAppSelector(selectors.corner);
  const items = useAppSelector(selectors.items);
  const values = useAppSelector(selectors.values);

  const dispatch = useAppDispatch();

  const addRow = () => {
    dispatch(actions.addRow());
  };

  const disableAddRow = items.some((item) => !item);

  return (
    <Table
      sx={{
        width: '100%',
        borderRight: (theme) => `1px solid ${theme.palette.divider}`,
      }}
    >
      <TableHead>
        <TableRow>
          <TableCorner>{corner}</TableCorner>
        </TableRow>
      </TableHead>
      <TableBody>
        {values.map((value, index) => {
          return (
            <Row
              key={value.id}
              selectors={selectors}
              actions={actions}
              index={index}
            />
          );
        })}

        <TableRow>
          <AddCell colSpan={2}>
            <TableActionButton
              buttonType="add"
              onClick={addRow}
              disabled={disableAddRow}
            />
          </AddCell>
        </TableRow>
      </TableBody>
    </Table>
  );
};

export default ItemSelect;

type RowProps = {
  selectors: RootSelectors;
  actions: any;
  index: number;
};

const Row: React.FC<RowProps> = React.memo(({ selectors, actions, index }) => {
  const dispatch = useAppDispatch();

  const itemSelectOptions = useAppSelector(selectors.itemSelectOptions);
  const items = useAppSelector(selectors.items);
  const value = useAppSelector(selectors.selectValueByIndex(index));

  const handleAutocompleteChange = (newValue: ItemSelectOption | null) => {
    if (!newValue) return;

    const savedValue =
      newValue.label === ADD_CUSTOM_ITEM_LABEL ? '' : newValue.label;

    dispatch(actions.setItemsOnIndex({ data: savedValue, index }));

    dispatch(
      actions.setValuesOnIndex({
        index,
        data: {
          value: newValue.value.toString(),
          id: value.id,
        },
      }),
    );

    if (!items.includes(newValue.label) && index === items.length - 1) {
      dispatch(actions.addRow());
    }
  };

  const handleInputChange = (value: string) => {
    dispatch(actions.setItemsOnIndex({ data: value, index }));
  };

  return (
    <TableRow>
      <TableCell>
        {value.value != '-1' ? (
          <Autocomplete
            value={itemSelectOptions.find(
              (option) => option.value.toString() == value.value,
            )}
            options={itemSelectOptions}
            groupBy={(option) => option.type}
            getOptionLabel={(option) => option.label}
            getOptionDisabled={(option) =>
              items.includes(option.label) &&
              option.label !== ADD_CUSTOM_ITEM_LABEL
            }
            renderInput={(params) => <TextField {...params} />}
            clearIcon={null}
            onChange={(_, value) => {
              handleAutocompleteChange(value);
            }}
            sx={{
              position: 'absolute',
              inset: 0,

              '.MuiOutlinedInput-root': {
                borderRadius: '0',
                height: '48px',
                paddingRight: '36px !important',

                fieldset: {
                  border: 'none',
                  height: '48px',
                },
              },

              input: {
                height: `${48 - 18}px !important`,
                padding: '0px !important',
              },
            }}
          />
        ) : (
          <TableInput
            placeholder="Zadajte názov položky..."
            inputTextAlign="left"
            defaultValue={items[index]}
            onBlur={(e) => handleInputChange(e.target.value)}
          />
        )}
      </TableCell>
    </TableRow>
  );
});
