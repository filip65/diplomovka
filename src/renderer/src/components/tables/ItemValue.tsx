import styled from '@emotion/styled';
import { useAppSelector } from '../../store/hooks';
import { RootSelectors } from '../../store/store';
import {
  Table,
  TableBody,
  TableCell,
  TableCorner,
  TableHead,
  TableRow,
  ActionCellBottom,
} from './Table';

const TableCellStyled = styled(TableCell)`
  padding: 0 9px;
  text-align: left;
`;

type ItemValueProps = {
  selectors: RootSelectors;
};

const ItemValue: React.FC<ItemValueProps> = ({ selectors }) => {
  const corner = useAppSelector(selectors.corner);
  const items = useAppSelector(selectors.items);

  return (
    <Table>
      <TableHead>
        <TableRow>
          <TableCorner>{corner}</TableCorner>
        </TableRow>
      </TableHead>
      <TableBody>
        {items.map((value: string, row: number) => {
          return (
            <TableRow key={row}>
              <TableCellStyled
                key={value + row.toString()}
                dangerouslySetInnerHTML={{ __html: value }}
              />
            </TableRow>
          );
        })}
        <TableRow>
          <ActionCellBottom></ActionCellBottom>
        </TableRow>
      </TableBody>
    </Table>
  );
};

export default ItemValue;
