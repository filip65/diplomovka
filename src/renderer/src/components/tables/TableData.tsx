import { Box, styled } from '@mui/material';

export const TableData: any = styled(Box)`
  overflow: auto;

  @media print {
    overflow: unset;
  }
`;
