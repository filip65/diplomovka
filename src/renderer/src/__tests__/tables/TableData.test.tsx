import '@testing-library/jest-dom';
import { TableData } from '@renderer/components/tables/TableData';
import { render } from '@testing-library/react';
import { matchers } from '@emotion/jest';

expect.extend(matchers);

describe('TableData styled component', () => {
  it('should have overflow: auto by default', () => {
    const { container } = render(<TableData>Content</TableData>);
    expect(container.firstChild).toHaveStyleRule('overflow', 'auto');
  });

  it('should set overflow to unset in print media query', () => {
    const { container } = render(<TableData>Content</TableData>);
    expect(container.firstChild).toHaveStyleRule('overflow', 'unset', {
      media: 'print',
    });
  });

  it('matches the snapshot', () => {
    const { container } = render(<TableData>Content</TableData>);
    expect(container.firstChild).toMatchSnapshot();
  });
});
