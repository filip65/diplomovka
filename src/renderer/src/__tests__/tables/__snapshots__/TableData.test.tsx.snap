// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableData styled component matches the snapshot 1`] = `
<div
  class="MuiBox-root css-9be8zc"
>
  Content
</div>
`;
